### PR TITLE
Simple Wallet Rust Transaction Processor

### PR DESCRIPTION
**Pull request by manojgop**
Added functionality for rust transaction processor for simple wallet
Modified rust TP dockerfile to avoid rebuilding dependencies on very build
Simple Wallet Rust TP docker build file to start containers

Signed-off-by: Manoj Gopalakrishnan <manoj.gopalkrishnan@intel.com>

### CODE REVIEW

File: rustprocessor/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+[package]
+name = "simplewallet-rust-tp"
+version = "0.8.0"
```
Comment (askmish) [inline rustprocessor/Cargo.toml]: Any reason for 0.8.0?

File: rustprocessor/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+authors = ["Intel Corporation."]
+description = "Sawtooth Simple Wallet Transaction Processor"
+
+[dependencies]
+#sawtooth_sdk = { path = "~/sawtooth/sawtooth-core/sdk/rust" }
```
Comment (askmish) [inline rustprocessor/Cargo.toml]: Please remove unnecessary comments from cargo.toml

File: rustprocessor/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# Copyright 2018 Bitwise IO, Inc.
```
Comment (askmish) [inline rustprocessor/Dockerfile]: Please fix the copyright header.

File: rustprocessor/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+WORKDIR /project/simplewallet/rustprocessor
+
+#CMD cd rustprocessor \
+CMD echo "\033[0;32m--- Building simplewallet-rust-tp ---\n\033[0m" \
+ && rm -rf ./bin/ \
+ && mkdir -p ./bin/ \
+ && cargo build --release \
+ && cp ./target/release/simplewallet-rust-tp ./bin/simplewallet-rust-tp \
+ && cargo run --release --bin simplewallet-rust-tp -- -v -C tcp://validator:4004 && tail -f /dev/null
```
Comment (askmish) [inline rustprocessor/Dockerfile]: Please split this into 80 chars

File: rustprocessor/src/handler/handler.rs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+use sawtooth_sdk::processor::handler::ApplyError;
+use sawtooth_sdk::processor::handler::TransactionContext;
+use sawtooth_sdk::processor::handler::TransactionHandler;
+
+use handler::payload::SwPayload;
+use handler::payload::Action;
+use handler::state::SwState;
+use handler::state::get_sw_prefix;
+
+
```
Comment (askmish) [inline rustprocessor/src/handler/handler.rs]: Extra blank line

File: rustprocessor/src/handler/handler.rs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+pub struct SwTransactionHandler {
+    family_name: String,
+    family_versions: Vec<String>,
```
Comment (askmish) [inline rustprocessor/src/handler/handler.rs]: Why is family_version a vector of strings?

Comment (manojgop) [reply rustprocessor/src/handler/handler.rs]: Rust SDK expects vector of Strings

File: rustprocessor/src/handler/mod.rs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2018 Intel Corportion.
```
Comment (danintel) [inline rustprocessor/src/handler/mod.rs]: s/Corportion/Corporation/

File: rustprocessor/src/handler/payload.rs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        let payload_string = match str::from_utf8(&payload_data) {
+            Ok(s) => s,
+            Err(_) => {
+                return Err(ApplyError::InvalidTransaction(String::from(
+                    "Invalid payload serialization",
+                )))
+            }
+        };
+
+        //SimpleWallet payload is constructed as comma seperated items
```
Comment (danintel) [inline rustprocessor/src/handler/payload.rs]: s/seperated/separated/

File: rustprocessor/src/handler/payload.rs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        let items: Vec<&str> = payload_string.split(",").collect();
+
+        if items.len() < 2 {
+            return Err(ApplyError::InvalidTransaction(String::from(
+                "Payload must have atleast 1 commas",
```
Comment (danintel) [inline rustprocessor/src/handler/payload.rs]: s/atleast 1 commas/at least 1 comma/

File: rustprocessor/src/handler/payload.rs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+            )));
+        }
+        
+        if items.len() > 3 {
+            return Err(ApplyError::InvalidTransaction(String::from(
+                "Payload must have atmost 2 commas",
```
Comment (danintel) [inline rustprocessor/src/handler/payload.rs]: s/atmost/at most/

File: rustprocessor/src/handler/payload.rs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                                )))
+                          }
+        };
+ 
+         
+        let value: u32 = match value.parse() {
+            Ok(num) => num,
+            Err(_) => {
+                return Err(ApplyError::InvalidTransaction(String::from(
+                    "Space must be an integer",
```
Comment (danintel) [inline rustprocessor/src/handler/payload.rs]: I don't understand this message: "Space must be an integer"
Do you mean the integer is missing (empty)? Or what?

**Review by danintel — CHANGES_REQUESTED** [review]
Minor comments about spelling an one error message.

File: rustprocessor/src/handler/mod.rs (hunk 7 above)
Comment (manojgop) [reply rustprocessor/src/handler/mod.rs]: ok

File: rustprocessor/src/handler/payload.rs (hunk 8 above)
Comment (manojgop) [reply rustprocessor/src/handler/payload.rs]: ok

File: rustprocessor/src/handler/payload.rs (hunk 9 above)
Comment (manojgop) [reply rustprocessor/src/handler/payload.rs]: ok

File: rustprocessor/src/handler/payload.rs (hunk 10 above)
Comment (manojgop) [reply rustprocessor/src/handler/payload.rs]: ok

File: rustprocessor/src/handler/payload.rs (hunk 11 above)
Comment (manojgop) [reply rustprocessor/src/handler/payload.rs]: changed message and submitted new commit

File: rustprocessor/src/handler/payload.rs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                )))
+            }
+        };
+
+        //SimpleWallet payload is constructed as comma separated items
+        let items: Vec<&str> = payload_string.split(",").collect();
+
+        if items.len() < 2 {
+            return Err(ApplyError::InvalidTransaction(String::from(
+                "Payload must have atleast 1 comma",
```
Comment (danintel) [inline rustprocessor/src/handler/payload.rs]: You still need to fix this:
s/atleast/at least/

**Review by danintel — CHANGES_REQUESTED** [review]
One fix still needs to be made.

File: rustprocessor/src/handler/payload.rs (hunk 12 above)
Comment (manojgop) [reply rustprocessor/src/handler/payload.rs]: ok

**Review by danintel — APPROVED** [review]

**Pull request merged**